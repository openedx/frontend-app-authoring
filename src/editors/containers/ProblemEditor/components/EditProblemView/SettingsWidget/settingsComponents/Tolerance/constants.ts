import messages from './messages';

export const ToleranceTypes = {
  percent: {
    type: 'Percent',
    message: messages.typesPercentage,
  },
  number: {
    type: 'Number',
    message: messages.typesNumber,
  },
  none: {
    type: 'None',
    message: messages.typesNone,
  },
} as const;
