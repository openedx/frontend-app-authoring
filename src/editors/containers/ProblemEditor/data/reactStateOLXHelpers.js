import { flatten } from 'lodash';

/**
 * flattenSubNodes - appends a nodes children to a given array 'subNodes' in a flattened format.
 * Only appends direct children unless `recursive` is set to true.
 * @param {object} node
 * @param {object[]} subNodes
 * @param {object} options - { recursive: false } Whether to search recursively for child tags.
 * @returns {void}
 */
function flattenSubNodes(node, subNodes, options = { recursive: false }) {
  const values = Array.isArray(node) ? (
    flatten(node.map(n => Object.values(n)))
  ) : Object.values(node);

  values.forEach(value => {
    if (Array.isArray(value)) {
      subNodes.push(...value);

      if (options.recursive) {
        value.forEach(nestedNode => {
          flattenSubNodes(nestedNode, subNodes, options);
        });
      }
    }
  });
}

/**
 * function nodeContainsChildTags - checks whether a node contains any subnodes with the specified tag names.
 *
 * @param {object} node - Example for node:
 *   {"div":
 *     [
 *       {"label":
 *         [
 *           {"#text":"def"}
 *         ]
 *       },
 *       {"div":
 *         [{
 *           label:
 *             [
 *               { '#text': 'def' },
 *             ],
 *         }],
 *       },
 *       {"#text":" "},
 *       {"em":
 *         [
 *           {"#text":"ghi"}
 *         ],
 *         ":@": {"@_class":"olx_description"}
 *       },
 *       {"em":
 *         [
 *           {"#text":"jkl"}
 *         ]
 *       }
 *     ]
 *   }
 * @param {string[]} tagNames
 * @param {boolean} [recursive=false] - Whether to search recursively for child tags.
 *
 * @returns {boolean} whether node contains the specified child tags
 */
export function nodeContainsChildTags(node, tagNames, options = { recursive: false }) {
  const subNodes = [];
  flattenSubNodes(node, subNodes, options);
  const res = subNodes.some(subNode => tagNames.includes(Object.keys(subNode)[0]));
  return res;
}

/**
 * @param {object} node
 * @returns {string} the tag name of the node
 */
export function tagName(node) {
  if (Array.isArray(node)) {
    throw new TypeError('function tagName does not accept arrays as input');
  }
  return Object.keys(node).find(key => key.match(/^[a-zA-Z].*/));
}

/**
 * findNodesAndRemoveTheirParentNodes - given 'arrayOfNodes', an array that results from the
 * XMLParser, this looks for 'nodesToFind': specific child tags (e.g. 'label'), and removes
 * their parent tags (e.g. 'div') but keeps all their content one level higher.
 * The 'parentsToRemove' array should contain a list of parent tags that will be removed.
 *
 * For example, if you have a div containing a p that contains a label and an em, and you specify
 * 'label' as the nodeToFind and 'p' as the parentToRemove, then you will get a div containing a label and an em.
 * @param {object} { arrayOfNodes, nodesToFind, parentsToRemove }
 *
 * @returns {object} - an array of nodes
 */
export function findNodesAndRemoveTheirParentNodes({
  arrayOfNodes,
  nodesToFind,
  parentsToRemove,
}) {
  const result = [];
  arrayOfNodes.forEach((node) => {
    const containsRelevantSubnodes = nodeContainsChildTags(node, nodesToFind);

    if (!containsRelevantSubnodes || !parentsToRemove.includes(tagName(node))) {
      result.push(node);
    } else {
      const subNodes = Object.values(node)[0];
      result.push(...subNodes);
    }
  });
  return result;
}
