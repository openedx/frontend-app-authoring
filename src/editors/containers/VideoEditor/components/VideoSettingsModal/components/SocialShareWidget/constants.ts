export const analyticsEvents = {
  socialSharingSettingChanged: 'edx.social.video_sharing_setting.changed',
};
