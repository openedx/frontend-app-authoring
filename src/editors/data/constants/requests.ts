import { StrictDict } from '../../utils';

export const RequestStates = StrictDict({
  inactive: 'inactive',
  pending: 'pending',
  completed: 'completed',
  failed: 'failed',
} as const);

export const RequestKeys = StrictDict({
  fetchVideos: 'fetchVideos',
  fetchBlock: 'fetchBlock',
  fetchImages: 'fetchImages',
  fetchUnit: 'fetchUnit',
  fetchStudioView: 'fetchStudioView',
  saveBlock: 'saveBlock',
  uploadVideo: 'uploadVideo',
  allowThumbnailUpload: 'allowThumbnailUpload',
  uploadThumbnail: 'uploadThumbnail',
  uploadTranscript: 'uploadTranscript',
  deleteTranscript: 'deleteTranscript',
  fetchCourseDetails: 'fetchCourseDetails',
  updateTranscriptLanguage: 'updateTranscriptLanguage',
  getTranscriptFile: 'getTranscriptFile',
  checkTranscriptsForImport: 'checkTranscriptsForImport',
  importTranscript: 'importTranscript',
  uploadAsset: 'uploadAsset',
  fetchAdvancedSettings: 'fetchAdvancedSettings',
  fetchVideoFeatures: 'fetchVideoFeatures',
} as const);
