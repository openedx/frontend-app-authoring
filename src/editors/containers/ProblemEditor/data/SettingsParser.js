import _ from 'lodash';

import { ShowAnswerTypes, RandomizationTypesKeys } from '../../../data/constants/problem';

export const popuplateItem = (parentObject, itemName, statekey, metadata, defaultValue = null, allowNull = false) => {
  let parent = parentObject;
  const item = _.get(metadata, itemName, null);

  // if item is null, undefined, or empty string, use defaultValue
  const finalValue = (!_.isNil(item) && item !== '') ? item : defaultValue;

  if (allowNull || (!_.isNil(finalValue) && finalValue !== defaultValue)) {
    parent = { ...parentObject, [statekey]: finalValue };
  }
  return parent;
};

export const parseScoringSettings = (metadata, defaultSettings) => {
  let scoring = {};

  const attempts = popuplateItem({}, 'max_attempts', 'number', metadata);
  const initialAttempts = _.get(attempts, 'number', null);
  const defaultAttempts = _.get(defaultSettings, 'max_attempts', null);
  attempts.unlimited = false;

  // isFinite checks if value is a finite primitive number.
  if (!_.isFinite(initialAttempts) || initialAttempts === defaultAttempts) {
    // set number to null in any case as lms will pick default value if it exists.
    attempts.number = null;
  }

  // if both block number and default number are null set unlimited to true.
  if (_.isNil(initialAttempts) && _.isNil(defaultAttempts)) {
    attempts.unlimited = true;
  }

  if (attempts.number < 0) {
    attempts.number = 0;
  }

  scoring = { ...scoring, attempts };

  scoring = popuplateItem(scoring, 'weight', 'weight', metadata);

  return scoring;
};

export const parseShowAnswer = (metadata) => {
  let showAnswer = {};

  const showAnswerType = _.get(metadata, 'showanswer', {});
  if (!_.isNil(showAnswerType) && showAnswerType in ShowAnswerTypes) {
    showAnswer = { ...showAnswer, on: showAnswerType };
  }

  showAnswer = popuplateItem(showAnswer, 'attempts_before_showanswer_button', 'afterAttempts', metadata);

  return showAnswer;
};

export const parseSettings = (metadata, defaultSettings) => {
  let settings = {};

  if (_.isNil(metadata) || _.isEmpty(metadata)) {
    return settings;
  }

  const scoring = parseScoringSettings(metadata, defaultSettings);
  if (!_.isEmpty(scoring)) {
    settings = { ...settings, scoring };
  }

  const showAnswer = parseShowAnswer(metadata);
  if (!_.isEmpty(showAnswer)) {
    settings = { ...settings, showAnswer };
  }

  const randomizationType = _.get(metadata, 'rerandomize', {});
  if (!_.isEmpty(randomizationType) && Object.values(RandomizationTypesKeys).includes(randomizationType)) {
    settings = popuplateItem(settings, 'rerandomize', 'randomization', metadata);
  }

  settings = popuplateItem(settings, 'show_reset_button', 'showResetButton', metadata);
  settings = popuplateItem(settings, 'submission_wait_seconds', 'timeBetween', metadata);
  return settings;
};
