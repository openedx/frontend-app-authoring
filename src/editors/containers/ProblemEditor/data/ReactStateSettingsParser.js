import { XMLParser } from 'fast-xml-parser';
import _ from 'lodash';

import {
  ShowAnswerTypesKeys,
} from '../../../data/constants/problem';
import { popuplateItem } from './SettingsParser';

const SETTING_KEYS = [
  'max_attempts',
  'weight',
  'showanswer',
  'show_reset_button',
  'rerandomize',
];

class ReactStateSettingsParser {
  constructor(problemState) {
    this.problem = problemState.problem;
    this.rawOLX = problemState.rawOLX;
  }

  getSettings() {
    let settings = {};
    const { defaultSettings } = this.problem;
    const stateSettings = this.problem.settings;

    const numberOfAttemptsChoice = [
      ShowAnswerTypesKeys.AFTER_SOME_NUMBER_OF_ATTEMPTS,
      ShowAnswerTypesKeys.AFTER_ALL_ATTEMPTS,
      ShowAnswerTypesKeys.AFTER_ALL_ATTEMPTS_OR_CORRECT,
    ];

    settings = popuplateItem(settings, 'number', 'max_attempts', stateSettings.scoring.attempts, defaultSettings?.maxAttempts, true);
    settings = popuplateItem(settings, 'weight', 'weight', stateSettings.scoring);
    settings = popuplateItem(settings, 'on', 'showanswer', stateSettings.showAnswer, defaultSettings?.showanswer, true);
    if (_.includes(numberOfAttemptsChoice, stateSettings.showAnswer.on)) {
      settings = popuplateItem(settings, 'afterAttempts', 'attempts_before_showanswer_button', stateSettings.showAnswer);
    }
    settings = popuplateItem(settings, 'showResetButton', 'show_reset_button', stateSettings, defaultSettings?.showResetButton, true);
    settings = popuplateItem(settings, 'timeBetween', 'submission_wait_seconds', stateSettings);
    settings = popuplateItem(settings, 'randomization', 'rerandomize', stateSettings, defaultSettings?.rerandomize, true);

    return settings;
  }

  parseRawOlxSettings() {
    const rawOlxSettings = this.getSettings();
    const parserOptions = {
      ignoreAttributes: false,
      alwaysCreateTextNode: true,
      numberParseOptions: {
        leadingZeros: false,
        hex: false,
      },
    };
    const parser = new XMLParser(parserOptions);
    const olx = parser.parse(this.rawOLX);
    const settingAttributes = Object.keys(olx.problem).filter(tag => tag.startsWith('@_'));
    settingAttributes.forEach(attribute => {
      const attributeKey = attribute.substring(2);
      if (SETTING_KEYS.includes(attributeKey)) {
        if (attributeKey === 'max_attempts' || attributeKey === 'weight') {
          rawOlxSettings[attributeKey] = parseInt(olx.problem[attribute], 10);
        } else {
          rawOlxSettings[attributeKey] = olx.problem[attribute];
        }
      }
    });
    return rawOlxSettings;
  }
}

export default ReactStateSettingsParser;
