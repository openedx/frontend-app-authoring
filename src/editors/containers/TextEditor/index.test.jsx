import 'CourseAuthoring/editors/setupEditorTest';
import React from 'react';
import { shallow } from '@edx/react-unit-test-utils';

import { formatMessage } from '../../testUtils';
import { actions, selectors } from '../../data/redux';
import { RequestKeys } from '../../data/constants/requests';
import { TextEditorInternal as TextEditor, mapStateToProps, mapDispatchToProps } from '.';

// Per https://github.com/tinymce/tinymce-react/issues/91 React unit testing in JSDOM is not supported by tinymce.
// Consequently, mock the Editor out.
jest.mock('@tinymce/tinymce-react', () => {
  const originalModule = jest.requireActual('@tinymce/tinymce-react');
  return {
    __esModule: true,
    ...originalModule,
    Editor: () => 'TiNYmCE EDitOR',
  };
});

jest.mock('../EditorContainer', () => 'EditorContainer');

jest.mock('./hooks', () => ({
  getContent: jest.fn(args => ({ getContent: args })),
  nullMethod: jest.fn().mockName('hooks.nullMethod'),
}));

jest.mock('../../sharedComponents/TinyMceWidget/hooks', () => ({
  ...jest.requireActual('../../sharedComponents/TinyMceWidget/hooks'),
  prepareEditorRef: jest.fn(() => ({
    editorRef: { current: { value: 'something' } },
    refReady: true,
    setEditorRef: jest.fn().mockName('hooks.prepareEditorRef.setEditorRef'),
  })),
}));

jest.mock('react', () => {
  const updateState = jest.fn();
  return {
    ...jest.requireActual('react'),
    updateState,
    useState: jest.fn(val => ([{ state: val }, jest.fn().mockName('setState')])),
  };
});

jest.mock('../../data/redux', () => ({
  __esModule: true,
  default: jest.fn(),
  actions: {
    app: {
      initializeEditor: jest.fn().mockName('actions.app.initializeEditor'),
    },
  },
  selectors: {
    app: {
      blockValue: jest.fn(state => ({ blockValue: state })),
      lmsEndpointUrl: jest.fn(state => ({ lmsEndpointUrl: state })),
      studioEndpointUrl: jest.fn(state => ({ studioEndpointUrl: state })),
      showRawEditor: jest.fn(state => ({ showRawEditor: state })),
      images: jest.fn(state => ({ images: state })),
      isLibrary: jest.fn(state => ({ isLibrary: state })),
      blockId: jest.fn(state => ({ blockId: state })),
      learningContextId: jest.fn(state => ({ learningContextId: state })),
    },
    requests: {
      isFailed: jest.fn((state, params) => ({ isFailed: { state, params } })),
      isFinished: jest.fn((state, params) => ({ isFailed: { state, params } })),
    },
  },
  thunkActions: {
    video: {
      importTranscript: jest.fn(),
    },
  },
}));

describe('TextEditor', () => {
  const props = {
    onClose: jest.fn().mockName('props.onClose'),
    // redux
    blockValue: { data: { data: 'eDiTablE Text' } },
    blockFailed: false,
    initializeEditor: jest.fn().mockName('args.intializeEditor'),
    showRawEditor: false,
    blockFinished: true,
    learningContextId: 'course+org+run',
    images: {},
    // inject
    intl: { formatMessage },
  };
  describe('snapshots', () => {
    test('renders as expected with default behavior', () => {
      expect(shallow(<TextEditor {...props} />).snapshot).toMatchSnapshot();
    });
    test('renders static images with relative paths', () => {
      const updatedProps = {
        ...props,
        blockValue: { data: { data: 'eDiTablE Text with <img src="/static/img.jpg" />' } },
      };
      expect(shallow(<TextEditor {...updatedProps} />).snapshot).toMatchSnapshot();
    });
    test('not yet loaded, Spinner appears', () => {
      expect(shallow(<TextEditor {...props} blockFinished={false} />).snapshot).toMatchSnapshot();
    });
    test('loaded, raw editor', () => {
      expect(shallow(<TextEditor {...props} showRawEditor />).snapshot).toMatchSnapshot();
    });
    test('block failed to load, Toast is shown', () => {
      expect(shallow(<TextEditor {...props} blockFailed />).snapshot).toMatchSnapshot();
    });
  });

  describe('mapStateToProps', () => {
    const testState = { A: 'pple', B: 'anana', C: 'ucumber' };
    test('blockValue from app.blockValue', () => {
      expect(
        mapStateToProps(testState).blockValue,
      ).toEqual(selectors.app.blockValue(testState));
    });
    test('blockFailed from requests.isFailed', () => {
      expect(
        mapStateToProps(testState).blockFailed,
      ).toEqual(selectors.requests.isFailed(testState, { requestKey: RequestKeys.fetchBlock }));
    });
    test('blockFinished from requests.isFinished', () => {
      expect(
        mapStateToProps(testState).blockFinished,
      ).toEqual(selectors.requests.isFinished(testState, { requestKey: RequestKeys.fetchBlock }));
    });
    test('learningContextId from app.learningContextId', () => {
      expect(
        mapStateToProps(testState).learningContextId,
      ).toEqual(selectors.app.learningContextId(testState));
    });
    test('images from app.images', () => {
      expect(
        mapStateToProps(testState).images,
      ).toEqual(selectors.app.images(testState));
    });
  });

  describe('mapDispatchToProps', () => {
    test('initializeEditor from actions.app.initializeEditor', () => {
      expect(mapDispatchToProps.initializeEditor).toEqual(actions.app.initializeEditor);
    });
  });
});
