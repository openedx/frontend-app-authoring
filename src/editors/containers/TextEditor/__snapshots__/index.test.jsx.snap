// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextEditor snapshots block failed to load, Toast is shown 1`] = `
<EditorContainer
  getContent={
    {
      "getContent": {
        "editorRef": {
          "current": {
            "value": "something",
          },
        },
        "showRawEditor": false,
      },
    }
  }
  isDirty={
    {
      "isDirty": {
        "editorRef": {
          "current": {
            "value": "something",
          },
        },
        "showRawEditor": false,
      },
    }
  }
  onClose={[MockFunction props.onClose]}
  returnFunction={null}
>
  <div
    className="editor-body h-75 overflow-auto"
  >
    <Toast
      onClose={[MockFunction hooks.nullMethod]}
      show={true}
    >
      Error: Could Not Load Text Content
    </Toast>
    <TinyMceWidget
      disabled={false}
      editorContentHtml="eDiTablE Text"
      editorRef={
        {
          "current": {
            "value": "something",
          },
        }
      }
      editorType="text"
      enableImageUpload={true}
      height="100%"
      id={null}
      images={{}}
      initializeEditor={[MockFunction args.intializeEditor]}
      isLibrary={null}
      learningContextId="course+org+run"
      lmsEndpointUrl=""
      minHeight={500}
      onChange={[Function]}
      setEditorRef={[MockFunction hooks.prepareEditorRef.setEditorRef]}
      studioEndpointUrl=""
    />
  </div>
</EditorContainer>
`;

exports[`TextEditor snapshots loaded, raw editor 1`] = `
<EditorContainer
  getContent={
    {
      "getContent": {
        "editorRef": {
          "current": {
            "value": "something",
          },
        },
        "showRawEditor": true,
      },
    }
  }
  isDirty={
    {
      "isDirty": {
        "editorRef": {
          "current": {
            "value": "something",
          },
        },
        "showRawEditor": true,
      },
    }
  }
  onClose={[MockFunction props.onClose]}
  returnFunction={null}
>
  <div
    className="editor-body h-75 overflow-auto"
  >
    <Toast
      onClose={[MockFunction hooks.nullMethod]}
      show={false}
    >
      Error: Could Not Load Text Content
    </Toast>
    <RawEditor
      content={
        {
          "data": {
            "data": "eDiTablE Text",
          },
        }
      }
      editorRef={
        {
          "current": {
            "value": "something",
          },
        }
      }
      lang="html"
    />
  </div>
</EditorContainer>
`;

exports[`TextEditor snapshots not yet loaded, Spinner appears 1`] = `
<EditorContainer
  getContent={
    {
      "getContent": {
        "editorRef": {
          "current": {
            "value": "something",
          },
        },
        "showRawEditor": false,
      },
    }
  }
  isDirty={
    {
      "isDirty": {
        "editorRef": {
          "current": {
            "value": "something",
          },
        },
        "showRawEditor": false,
      },
    }
  }
  onClose={[MockFunction props.onClose]}
  returnFunction={null}
>
  <div
    className="editor-body h-75 overflow-auto"
  >
    <Toast
      onClose={[MockFunction hooks.nullMethod]}
      show={false}
    >
      Error: Could Not Load Text Content
    </Toast>
    <div
      className="text-center p-6"
    >
      <Spinner
        animation="border"
        className="m-3"
        screenreadertext="loading"
      />
    </div>
  </div>
</EditorContainer>
`;

exports[`TextEditor snapshots renders as expected with default behavior 1`] = `
<EditorContainer
  getContent={
    {
      "getContent": {
        "editorRef": {
          "current": {
            "value": "something",
          },
        },
        "showRawEditor": false,
      },
    }
  }
  isDirty={
    {
      "isDirty": {
        "editorRef": {
          "current": {
            "value": "something",
          },
        },
        "showRawEditor": false,
      },
    }
  }
  onClose={[MockFunction props.onClose]}
  returnFunction={null}
>
  <div
    className="editor-body h-75 overflow-auto"
  >
    <Toast
      onClose={[MockFunction hooks.nullMethod]}
      show={false}
    >
      Error: Could Not Load Text Content
    </Toast>
    <TinyMceWidget
      disabled={false}
      editorContentHtml="eDiTablE Text"
      editorRef={
        {
          "current": {
            "value": "something",
          },
        }
      }
      editorType="text"
      enableImageUpload={true}
      height="100%"
      id={null}
      images={{}}
      initializeEditor={[MockFunction args.intializeEditor]}
      isLibrary={null}
      learningContextId="course+org+run"
      lmsEndpointUrl=""
      minHeight={500}
      onChange={[Function]}
      setEditorRef={[MockFunction hooks.prepareEditorRef.setEditorRef]}
      studioEndpointUrl=""
    />
  </div>
</EditorContainer>
`;

exports[`TextEditor snapshots renders static images with relative paths 1`] = `
<EditorContainer
  getContent={
    {
      "getContent": {
        "editorRef": {
          "current": {
            "value": "something",
          },
        },
        "showRawEditor": false,
      },
    }
  }
  isDirty={
    {
      "isDirty": {
        "editorRef": {
          "current": {
            "value": "something",
          },
        },
        "showRawEditor": false,
      },
    }
  }
  onClose={[MockFunction props.onClose]}
  returnFunction={null}
>
  <div
    className="editor-body h-75 overflow-auto"
  >
    <Toast
      onClose={[MockFunction hooks.nullMethod]}
      show={false}
    >
      Error: Could Not Load Text Content
    </Toast>
    <TinyMceWidget
      disabled={false}
      editorContentHtml="eDiTablE Text with <img src="/asset+org+run+type@asset+block@img.jpg" />"
      editorRef={
        {
          "current": {
            "value": "something",
          },
        }
      }
      editorType="text"
      enableImageUpload={true}
      height="100%"
      id={null}
      images={{}}
      initializeEditor={[MockFunction args.intializeEditor]}
      isLibrary={null}
      learningContextId="course+org+run"
      lmsEndpointUrl=""
      minHeight={500}
      onChange={[Function]}
      setEditorRef={[MockFunction hooks.prepareEditorRef.setEditorRef]}
      studioEndpointUrl=""
    />
  </div>
</EditorContainer>
`;
